File: notifications/src/main/java/io/castled/notifications/trigger/DateConditionEvaluator.java
package io.castled.notifications.trigger;

import io.castled.notifications.exceptions.CastledRuntimeException;
import io.castled.notifications.trigger.models.BetweenOperation;
import io.castled.notifications.trigger.models.PropertyOperation;
import io.castled.notifications.trigger.models.SingleValueOperation;

public class DateConditionEvaluator implements ParamsConditionEvaluator {

    @Override
    public boolean evaluateCondition(Object value, PropertyOperation propertyOperation) {

        String dateText = (String) value;
        switch (propertyOperation.getOperationType()) {
            case BETWEEN:
                BetweenOperation dateInterval = (BetweenOperation) propertyOperation;
                return dateText != null && dateInterval.getFrom().compareTo(dateText) < 0 && dateInterval.getTo().compareTo(dateText) > 0;

            case EQ:
                SingleValueOperation dateEquals = (SingleValueOperation) propertyOperation;
                return dateEquals.getValue().equals(dateText);

            case NEQ:
                if (dateText == null) {
                    return true;
                }
                SingleValueOperation dateNotEquals = (SingleValueOperation) propertyOperation;
                return !dateNotEquals.getValue().equals(dateText);

            default:
                throw new CastledRuntimeException(String.format("Operations type %s not supported for timestamp operand", propertyOperation.getPropertyType()));
        }
    }
}

File: notifications/src/main/java/io/castled/notifications/trigger/DateConditionEvaluator.kt
package io.castled.notifications.trigger

import io.castled.notifications.trigger.enums.OperationType
import io.castled.notifications.trigger.models.BetweenOperation
import io.castled.notifications.trigger.models.PropertyOperation
import io.castled.notifications.trigger.models.SingleValueOperation

internal class DateConditionEvaluator : ParamsConditionEvaluator() {

    // TODO: Need to fix the date evaluation
    override fun evaluateCondition(value: Any?, propertyOperation: PropertyOperation): Boolean {
        val dateText = value as String?
        return when (propertyOperation.type) {
            OperationType.BETWEEN -> {
                val dateInterval = propertyOperation as BetweenOperation
                dateText != null && dateInterval.from < dateText && dateInterval.to > dateText
            }
            OperationType.EQ -> {
                val dateEquals = propertyOperation as SingleValueOperation
                dateEquals.value == dateText
            }
            OperationType.NEQ -> {
                if (dateText == null) {
                    return true
                }
                val dateNotEquals = propertyOperation as SingleValueOperation
                dateNotEquals.value != dateText
            }
            else -> {
                logger.error("Operations type ${propertyOperation.propertyType} not supported for date operand")
                return false
            }
        }
    }
}